File: 242. Valid Anagram/Solution.java
// LeetCode #242: Valid Anagram

/*
Given two strings s and t, write a function to determine if t is an anagram of s.

For example,
s = "anagram", t = "nagaram", return true.
s = "rat", t = "car", return false.

Note:
You may assume the string contains only lowercase alphabets.
*/

/*
public class Solution {
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) return false;
        
        Map<Character, Integer> map = new HashMap<Character, Integer>();
        for (int i = 0; i < s.length(); i++) {
            if (map.containsKey(s.charAt(i)))
                map.put(s.charAt(i), map.get(s.charAt(i)) + 1); // cannot find symbol: method set(char,int)
            else
                map.put(s.charAt(i), 1);
        }
        
        for (int i = 0; i < t.length(); i++) {
            if (!map.containsKey(t.charAt(i)))
                return false;
            else {
                if (map.get(t.charAt(i)) == 0)
                    return false;
                else
                    map.put(t.charAt(i), map.get(t.charAt(i)) - 1);
            }
        }
        return true;
    }
}
*/

public class Solution {
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) return false;
        
        char[] sarr = s.toCharArray();
        char[] tarr = t.toCharArray();
        
        Arrays.sort(sarr);
        Arrays.sort(tarr);
        
        return String.valueOf(sarr).equals(String.valueOf(tarr));
    }
}

File: 242. Valid Anagram/Solution.cpp
// LeetCode #242: Valid Anagram

/*
Given two strings s and t, write a function to determine if t is an anagram of s.

For example,
s = "anagram", t = "nagaram", return true.
s = "rat", t = "car", return false.

Note:
You may assume the string contains only lowercase alphabets.
*/

class Solution {
public:
    bool isAnagram(string s, string t) {
        if (s.size() != t.size()) return false;
        unordered_map<char, int> map;
        
        for (int i = 0; i < s.size(); i++)
            if (map.find(s[i]) == map.end())
                map[s[i]] = 1;
            else
                map[s[i]]++;
            
        for (int i = 0; i < t.size(); i++)
            if (map.find(t[i]) == map.end())
                return false;
            else {
                if (map[t[i]] >= 1)
                    map[t[i]]--;
                else
                    return false;
            }
            
        return true;
    }
};

/*
class Solution {
public:
    bool isAnagram(string s, string t) {
        sort(s.begin(), s.end());
        sort(t.begin(), t.end());
        return s == t;
    }
};
*/